compio/benches/fs.rs

```diff
@@ -1,92 +1,609 @@
-use criterion::{criterion_group, criterion_main, Criterion};
+use std::{
+    hint::black_box,
+    io::{Read, Seek, SeekFrom, Write},
+    path::Path,
+    time::Instant,
+};
+
+use compio_buf::{IntoInner, IoBuf};
+use compio_io::{AsyncReadAt, AsyncWriteAt};
+use criterion::{criterion_group, criterion_main, Bencher, Criterion};
+use futures_util::{stream::FuturesUnordered, StreamExt};
+use rand::{thread_rng, Rng, RngCore};
 use tempfile::NamedTempFile;
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt};
+
+#[cfg(target_os = "linux")]
+struct MonoioRuntime(std::cell::RefCell<monoio::Runtime<monoio::IoUringDriver>>);
+
+#[cfg(target_os = "linux")]
+impl criterion::async_executor::AsyncExecutor for MonoioRuntime {
+    fn block_on<T>(&self, future: impl futures_util::Future<Output = T>) -> T {
+        self.0.borrow_mut().block_on(future)
+    }
+}
+
+#[cfg(target_os = "linux")]
+impl criterion::async_executor::AsyncExecutor for &MonoioRuntime {
+    fn block_on<T>(&self, future: impl futures_util::Future<Output = T>) -> T {
+        self.0.borrow_mut().block_on(future)
+    }
+}
 
 criterion_group!(fs, read, write);
 criterion_main!(fs);
 
+const BUFFER_SIZE: usize = 4096;
+
+fn read_std(b: &mut Bencher, (path, offsets): &(&Path, &[u64])) {
+    let file = std::fs::File::open(path).unwrap();
+    b.iter(|| {
+        let mut buffer = [0u8; BUFFER_SIZE];
+        for &offset in *offsets {
+            #[cfg(windows)]
+            {
+                use std::os::windows::fs::FileExt;
+                file.seek_read(&mut buffer, offset).unwrap();
+            }
+            #[cfg(unix)]
+            {
+                use std::os::unix::fs::FileExt;
+                file.read_at(&mut buffer, offset).unwrap();
+            }
+        }
+        buffer
+    })
+}
+
+fn read_tokio(b: &mut Bencher, (path, offsets): &(&Path, &[u64])) {
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+    b.to_async(&runtime).iter_custom(|iter| async move {
+        let mut file = tokio::fs::File::open(path).await.unwrap();
+
+        let start = Instant::now();
+        for _i in 0..iter {
+            let mut buffer = [0u8; BUFFER_SIZE];
+            for &offset in *offsets {
+                file.seek(SeekFrom::Start(offset)).await.unwrap();
+                _ = file.read(&mut buffer).await.unwrap();
```

Comment:
Please calculate the read size to determine its throughput

Reply:
AFAIK it's hard to do so in criterion.

Reply:
Add `group.throughput(criterion::Throughput::Bytes(16 * 1024 * 1024));` for the entire file maybe enough?

Reply:
It's for fixed throughput. Here the throughput may differ by functions.

Reply:
I've added throughput for read-all and write-all benches.